protomap/proto.go

```diff
@@ -700,6 +726,137 @@ func protoFromPathsInternal(p proto.Message, vals map[*gpb.Path]any, valPrefix,
 	return nil
 }
 
+// createListField creates the entries of the repeated field fd within the protobuf message m, mapping the values within the val map.
+// valPrefix specifies the prefix to be applied to the paths within the vals map, protoPrefix specifies the prefix for the protobuf
+// message (if it is not the root), and fieldPath specifies the path to the field that is being mapped. ignoreExtras indicates whether
+// extra paths that do not exist in the message should be treated as errors.
+func createListField(m proto.Message, fd protoreflect.FieldDescriptor, fieldPath, valPrefix, protoPrefix *gpb.Path, vals map[*gpb.Path]any, ignoreExtras bool) (protoreflect.Value, error) {
+	keys := []map[string]string{}
+	keyPaths := []*gpb.Path{}
+
+	// We need to identify the keys that are within the list, as well as the data tree paths
+	// that they correspond to. We walk through the supplied values to determine which to process.
+	for p := range vals {
+		// Make the paths within the vals map absolute according to the supplied valPrefix.
+		absPath := &gpb.Path{
+			Elem: append(append([]*gpb.PathElem{}, valPrefix.Elem...), p.Elem...),
+		}
+		// Since the fieldPath is a schema path, then we need to compare just schema paths
+		// to avoid comparing the keys.
+		if !util.PathMatchesPathElemPrefix(schemaPath(absPath), schemaPath(fieldPath)) {
+			continue
+		}
+		// The key of the list is in the last element of the absolute path in the values map (the values
+		// map MUST contain data tree paths, since it is telling us list values to unmarshal).
+		k := absPath.Elem[len(fieldPath.Elem)-1]
+		// If the last element doesn't have a key, then we have not correctly found the list.
+		if k.Key == nil {
+			return protoreflect.Value{}, fmt.Errorf("invalid list schema field path does not have a list at the end %s", fieldPath)
+		}
+
+		// Find the parts of the path that are not the list -- we assume that this is 2 elements since
+		// we are in a compressed schema.
+		// TODO(robjs): Currently, this may report incorrectly in an uncompressed schema, but we don't have
+		// a signal to indicate this. One needs to be added to the generated protobufs.
+		keyPath := &gpb.Path{
+			Elem: append(protoPrefix.Elem, absPath.Elem[len(protoPrefix.Elem):len(protoPrefix.Elem)+2]...),
```

Comment:
suggestion to append from an empty slice.

I'm concerned this might not always work because if `append` here doesn't always allocate a new array then there might be conflicts between different iterations of this.

Reply:
Good point -- I had forgotten that there are cases where `append` will not allocate a new array. I've made the change.